framework/src/modules/chain/bft/finality_manager.js

```diff
@@ -0,0 +1,297 @@
+		const maxPreCommitHeight = header.height - 1;
+
+		for (let j = minPreCommit; j <= maxPreCommitHeight; j++) {
+			// Add pre-commit if threshold is reached
+			if (this.preVotes[j] >= this.preVoteThreshold) {
+				// Increase the pre-commit for particular height
+				this.preCommits[j] = (this.preCommits[j] || 0) + 1;
+
+				// Keep track of the last pre-commit point
+				delegateState.maxPreCommitHeight = j;
+			}
+		}
+
+		// Check between height of first block of the round when delegate was active
+		// Or one step ahead where it forged the last block
+		// Or one step ahead where it left the last pre-vote
+		// Or maximum 3 rounds backward
+		const minPreVoteHeight = Math.max(
+			delegateMinHeightActive,
+			header.maxHeightPreviouslyForged + 1,
+			delegateState.maxPreVoteHeight + 1,
+			header.height - this.processingThreshold
+		);
+		// Pre-vote upto current block height
+		const maxPreVoteHeight = header.height;
+		for (let j = minPreVoteHeight; j <= maxPreVoteHeight; j++) {
+			this.preVotes[j] = (this.preVotes[j] || 0) + 1;
+		}
+		// Update delegate state
+		delegateState.maxPreVoteHeight = maxPreVoteHeight;
+
+		// Set the delegate state
+		this.state[delegatePublicKey] = delegateState;
+	}
+
+	/**
+	 * Update the pre-voted confirmed and finalized height
+	 */
+	updatePreVotedAndFinalizedHeight() {
+		debug('updatePreVotedAndFinalizedHeight invoked');
+		if (this.headers.length === 0) {
+			return;
+		}
+
+		const higherPairVoted = Object.entries(this.preVotes)
+			.reverse()
+			.find(pair => pair[1] >= this.preVoteThreshold);
+
+		this.prevotedConfirmedHeight = higherPairVoted
+			? parseInt(higherPairVoted[0])
+			: this.prevotedConfirmedHeight;
+
+		const higherPairCommitted = Object.entries(this.preCommits)
+			.reverse()
+			.find(pair => pair[1] >= this.preCommitThreshold);
+
+		this.finalizedHeight = higherPairCommitted
+			? parseInt(higherPairCommitted[0])
+			: this.finalizedHeight;
+	}
+
+	/**
+	 * Use existing block headers and re-compute all information
+	 */
+	recompute() {
+		this.state = {};
+		this.finalizedHeight = 0;
+		this.prevotedConfirmedHeight = 0;
+		this.preVotes = {};
+		this.preCommits = {};
+
+		this.headers.items.forEach(header => {
+			this.updatePreVotesPreCommits(header);
+			this.updatePreVotedAndFinalizedHeight();
+		});
+	}
+
+	/**
+	 * Verify if the block header is good for current chain
+	 *
+	 * @param {BlockHeader} blockHeader
+	 */
+	verifyBlockHeaders(blockHeader) {
+		debug('verifyBlockHeaders invoked');
+		// We need minimum processingThreshold to decide
+		// if prevotedConfirmedUptoHeight is correct
+		if (
+			this.headers.length >= this.processingThreshold &&
+			blockHeader.prevotedConfirmedUptoHeight !== this.prevotedConfirmedHeight
+		) {
+			throw new BFTInvalidAttributeError(
+				'Wrong prevotedConfirmedHeight in blockHeader.'
+			);
+		}
+
+		// Find top most block forged by same delegate
+		const delegateLastBlock = this.headers
+			.top(this.processingThreshold)
```

Comment:
I couldn't fine proper documentation for array.top. Is it a standart array method?

Reply:
Its the method from `HeadsList`